src/document/delta/transaction-base.js

```diff
@@ -0,0 +1,139 @@
+		 */
+		constructor( doc ) {
+			/**
+			 * Document which this transaction changes.
+			 *
+			 * @readonly
+			 * @type {document.Document}
+			 */
+			this.doc = doc;
+
+			/**
+			 * Array of deltas which compose transaction.
+			 *
+			 * @readonly
+			 * @type {document.delta.Delta[]}
+			 */
+			this.deltas = [];
+		}
+
+		/**
+		 * Adds delta to the transaction instance. All modification methods (insert, remove, split, etc.) use this method
+		 * to add created deltas.
+		 *
+		 * @param {document.delta.Delta} delta Delta to add.
+		 *
+		 * @return {document.delta.Delta} Added delta.
+		 */
+		addDelta( delta ) {
+			delta.transaction = this;
+			this.deltas.push( delta );
+
+			return delta;
+		}
+
+		/**
+		 * Transaction provides iterator interface which will iterate over deltas in the transaction.
+		 */
+		[ Symbol.iterator ]() {
+			return this.deltas[ Symbol.iterator ]();
+		}
+
+		/**
+		 * Static method to register transaction methods. To make code scalable transaction do not have modification
+		 * methods build in. They can be registered using this method.
+		 *
+		 * This method check if there is no naming collision and throw `transaction-register-taken` if the method name
+		 * is already taken.
+		 *
+		 * It also pass {@link document.Document} and {@link document.Transaction} do the creator class.
+		 *
+		 * Registered function returns `this` so they because chainable by default.
+		 *
+		 * Beside that no magic happens here, the methods is added to the Transaction prototype.
+		 *
+		 * For example:
+		 *
+		 *		Transaction.register( 'insert', ( doc, transaction, position, nodes ) => {
+		 *			// You can use a class inherit from Delta if that class should handle OT in the special way.
+		 *			const delta = new Delta();
+		 *
+		 * 			// Create operations which should be components of this delta.
+		 *			const operation = new InsertOperation( position, nodes, doc.version );
+		 *
+		 *			// Remember to apply every operation, no magic, you need to do it manually.
+		 *			doc.applyOperation( operation );
+		 *
+		 *			// Add operation to the delta.
+		 *			delta.addOperation( operation );
+		 *
+		 *			// Add delta to the transaction instance.
+		 *			transaction.addDelta( delta );
+		 *
+		 * 			// You do not need to return transaction, register method will take care to make the method chainable.
+		 *		} );
+		 *
+		 * @param {String} name Method name.
+		 * @param {Fuction} creator Method body.
+		 */
+		static register( name, creator ) {
+			if ( Transaction.prototype[ name ] ) {
+				/**
+				 * This transaction method is already taken.
+				 *
+				 * @error transaction-register-taken
+				 * @param {String} name
+				 */
+				throw new CKEditorError(
+					'transaction-register-taken: This transaction method is already taken.',
+					{ name: name } );
+			}
+
+			Transaction.prototype[ name ] = function() {
+				creator.apply( this, [ this.doc, this ].concat( Array.from( arguments ) ) );
```

Comment:
I don't like that instead of assigning the creator directly to the prototype we wrap it and concatenate args with what would already be accessible as `this`. This is an unnecessary step.